### PR TITLE
[PartyMode] TicTacToe - improvements

### LINKED ISSUE #401: Odd number of players not working in Tic-Tac-Toe Party Mode
If you setup a Tic-Tac-Toe game (Latest nightly build as of 29.10.18) with two teams and for example 5 players (3 on team 1 and 2 on team 2), the game will only use the first 2 players of team 1 and match them against both of team 2. 
The 3rd player of the first team will never be chosen to sing a song.
For now I've just added a "Guest" to the second team, 3on3 is working just fine. But it should be possible to also play with an odd number of players.

### PR DESCRIPTION
**Pull request by damienlaguerre**
### Summary
Try to fix somes mistakes in the game mode.

### What does this PR do?

- Fix: Teams with more than 2 players. Only the first and the second players are chosen.
- Add: The number of jockers can be configured.
- Improve: The player random selection.

### Screenshots / Video
Capture d’écran 2023-12-22 173912

Capture d’écran 2023-12-22 173720

### CODE REVIEW

**Comment by marwin89** [comment]
Hi Damien, welcome to vocaluxe! this looks like a good contribution. 👋 :)
Can someone do a short QA and review? (@Stefan1200-de) I'm overloaded with other tasks (UI/Logo design, etc.) right now.

File: PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeMain.cs [hunk 1]
```diff
@@ -519,6 +522,56 @@ private void _StartPreview(int songID)
...
+                return random.Next(max+1);
+            }
+        }
+
+        private void _ChangePlayer(int teamNr = -1, bool next = false)
+        {
+
+            if (_PartyMode.GameData.ProfileIDsTeam1.Count > 0 && (teamNr == 0 || teamNr == -1))
+            {
+                if (next == true)
```
Comment (flokuep) [inline PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeMain.cs]: could be simplyfied to if (next)

File: PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeMain.cs [hunk 2]
```diff
@@ -519,6 +522,56 @@ private void _StartPreview(int songID)
             CBase.BackgroundMusic.LoadPreview(song);
         }
 
+        private int _GetRamdom(int max)
+        {
+            lock (syncLock)
```
Comment (flokuep) [inline PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeMain.cs:527]: Could you explain why a lock is necessary here?

File: PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeMain.cs [hunk 3]
```diff
@@ -519,6 +522,56 @@ private void _StartPreview(int songID)
...
+            { 
+                return random.Next(max+1);
+            }
+        }
+
+        private void _ChangePlayer(int teamNr = -1, bool next = false)
+        {
+
+            if (_PartyMode.GameData.ProfileIDsTeam1.Count > 0 && (teamNr == 0 || teamNr == -1))
```
Comment (flokuep) [inline PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeMain.cs:536]: I'd suggest to extract some code here to reduce duplicated lines, I guess a new method with Profile-TeamCount, teamNr and next would be working. Return value is ne id of the next player.

File: PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeMain.cs [hunk 4]
```diff
@@ -78,6 +81,12 @@ protected override int _ScreenVersion
         private const int _FieldSpace = 10;
         private float _FieldSize = 100;
 
+        int idPlayer1 = 0;
```
Comment (flokuep) [inline PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeMain.cs]: Should be private fields I guess?

File: PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeConfig.cs [hunk 5]
```diff
@@ -131,14 +134,27 @@ private void _FillSlides()
 
             _SelectSlides[_SelectSlideNumFields].SelectedTag = _PartyMode.GameData.NumFields;
 
+            _SelectSlides[_SelectSlideNumJockers].Clear();
+            _SelectSlides[_SelectSlideNumJockers].AddValue(0);
+            _SelectSlides[_SelectSlideNumJockers].AddValue(1);
+            _SelectSlides[_SelectSlideNumJockers].AddValue(2);
+            _SelectSlides[_SelectSlideNumJockers].AddValue(3);
```
Comment (flokuep) [inline PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeConfig.cs]: Could be improved by using a for-loop (or maybe AddValues can be used, not totally sure about it).

File: PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeNames.cs [hunk 6]
```diff
@@ -76,11 +77,9 @@ public override void Next()
                 _PartyMode.GameData.ProfileIDsTeam1 = _TeamList[0];
                 _PartyMode.GameData.ProfileIDsTeam2 = _TeamList[1];
             }
-            if (_NumPlayerTeams.Length == 2)
-            {
-                _PartyMode.GameData.NumPlayerTeam1 = _NumPlayerTeams[0];
-                _PartyMode.GameData.NumPlayerTeam2 = _NumPlayerTeams[1];
-            }
+            _PartyMode.GameData.NumPlayerTeam1 = _PartyMode.GameData.ProfileIDsTeam1.Count();
```
Comment (flokuep) [inline PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeNames.cs]: There is a Count property that could be used instead of the method I guess.

**Review by flokuep — CHANGES_REQUESTED** [review]
Thank you very much for your (first) contribution :) I really like the improvements you've added to that party mode. I've found some smaller things that can be improved within you code, I'd be happy if you could have a look on it (most of these things are also mentioned within the sonarqube report).

**Comment by damienlaguerre** [comment]
Thanks for your messages! 
I've modified the code thanks to your feedback.
I hope you like it.

File: PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeMain.cs (hunk 2 above)
Comment (damienlaguerre) [reply PartyModes/PartyModeTicTacToe/CPartyScreenTicTacToeMain.cs:527]: Without this, Random isn't really effective. It returns the same number for both teams at the start of the round.
If you know of a better way to generate a random int I'd love to hear from you :)

**Review by flokuep — APPROVED** [review]

**Comment by flokuep** [comment]
Yes, I like them :) Sonar ist also mentioning the random-problematic, so that's fine I guess. PR is merged and changes will be included in the next nightly build (within a few minutes)

**Pull request merged**